File: House Robber/Mine.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int len = nums.size();
        if (len == 0)
            return 0;
        if (len == 1)
            return nums[0];
        if (len == 2)
            return max(nums[0], nums[1]);
        int result = INT_MIN;
        vector<int> dp(len, 0);
        vector<bool> flag (len, false);
        dp[0] = nums[0];
        flag[0] = true;
        if (nums[1] > nums[0]){
            dp[1] = nums[1];
            flag[1] = true;
        }
        else{
            dp[1] = nums[0];
            flag[1] = false;
        }
        for (int i = 2; i < len; i++){
            if (flag[i-1])
                if (dp[i-1] > dp[i-2] + nums[i]){
                    dp[i] = dp[i-1];
                    flag[i] = false;
                }
                else{
                    dp[i] = dp[i-2] + nums[i];
                    flag[i] = true;
                }
            else{
                dp[i] = dp[i-2] + nums[i];
                flag[i] = true;
            }
                
        }
        return dp[len - 1];
    }
};

File: 18. 4Sum/mine.py
class Solution:
    def twoSum(self, nums, num_a, num_b, num_c, num_d, target, result):
        while num_c < num_d:
            value = nums[num_c] + nums[num_d]
            if value > target:
                num_d -= 1
            elif value < target:
                num_c += 1
            else:
                result.append([nums[num_a], nums[num_b], nums[num_c], nums[num_d]])
                num_c = self.pass_repeated_number(nums, num_c, num_d, "LEFT")
                num_d = self.pass_repeated_number(nums, num_c, num_d, "RIGHT")
                num_c, num_d = num_c + 1, num_d - 1
    
    def pass_repeated_number(self, nums, left, right, side):
        if side == "LEFT":
            while left < right and nums[left] == nums[left + 1]:
                left += 1
            return left
        while right > left and nums[right] == nums[right - 1]:
            right -= 1
        return right
    
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if not nums or len(nums) < 4:
            return []
        
        nums.sort()
        result = []
        num_a, num_b = 0, 1
        while num_a < len(nums) - 3:
            while num_b < len(nums) - 2:
                two_target = target - nums[num_a] - nums[num_b]
                self.twoSum(nums, num_a, num_b, num_b + 1, len(nums) - 1, two_target, result)
                num_b = self.pass_repeated_number(nums, num_b, len(nums) - 2, "LEFT")
                num_b += 1
            num_a = self.pass_repeated_number(nums, num_a, len(nums) - 3, "LEFT")
            num_a, num_b = num_a + 1, num_a + 2
            
        return result
                    
                    

                
                
                
                        
                
        
